src/ast/optimize.ts

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) 2021 the Skulpt Project
+// SPDX-License-Identifier: MIT
+
+import type { AST, BinOp, BoolOp, cmpop, Compare, expr, mod, Subscript, Tuple, UnaryOp } from "./astnodes.ts";
+import {
+    Add,
+    And,
+    ASTVisitor,
+    ASTKind,
+    BitAnd,
+    BitOr,
+    BitXor,
+    Constant,
+    Div,
+    FloorDiv,
+    In,
+    Invert,
+    Is,
+    IsNot,
+    Load,
+    LShift,
+    Mod,
+    Not,
+    NotIn,
+    Mult,
+    Pow,
+    RShift,
+    Sub,
+    UAdd,
+    USub,
+} from "./astnodes.ts";
+import { pyFalse, pyTrue, pyTuple } from "../mock_types/constants.ts";
+import type { pyConstant } from "../mock_types/constants.ts";
+import {
+    pyAdd,
+    pyAnd,
+    pyDiv,
+    pyFloorDiv,
+    pyGetItem,
+    pyInvert,
+    pyIsTrue,
+    pyLshift,
+    pyMod,
+    pyMult,
+    pyNeg,
+    pyOr,
+    pyPos,
+    pyPow,
+    pyRshift,
+    pySub,
+    pyXor,
+} from "../mock_types/operator.ts";
+
+export function astOptimize(node: mod) {
+    const visitor = new OptimizingVisitor();
+    return node.mutateOver(visitor);
+}
+
+function asConstantTruth(node: AST) {
+    if (node._kind !== ASTKind.Constant) {
+        return null;
+    }
+    return pyIsTrue((node as Constant).value);
+}
+
+function makeConst(node: expr, newval: pyConstant | null) {
+    if (newval === null) {
+        return node;
+    }
+    return new Constant(newval, null, node.lineno, node.col_offset, node.end_lineno, node.end_col_offset);
+}
+
+function unaryNot(constant: pyConstant) {
+    return pyIsTrue(constant) ? pyFalse : pyTrue;
+}
+
+function makeConstTuple(elts: expr[]): null | pyTuple {
```

Comment:
It turns a `Tuple` ast node into a `Constant` ast node
Which I guess means a slightly faster compilation.
add a comment?

Reply:
only if all the elements are constants? this would construct a tuple compile time instead of calling it run time. I don't think that makes a difference from a skulpt perspective currently because those things won't get loaded until we hit that code right?

Reply:
Yeah it makes little difference for current skulpt implementation of creating constant tuples.

For skulpt it just means simplifying this code
https://github.com/skulpt/skulpt/blob/master/src/compile.js#L371-L394
since we've done the work here to check for constants - so no need to do it again.

Reply:
> Some things I thought about while working on this PR:
> 
> Maybe we should get rid of `Add`, `Load` and friends as instances of `AST` subclasses.
> https://github.com/mozillazg/pypy/blob/master/pypy/interpreter/astcompiler/ast.py#L2831
> 
> ```ts
> export const enum expr_context {
>     Load
> }
> export const Load = expr_context.Load;
> ```

This makes sense to me esp since in most places we wouldn't need to reflect on them only in the dump function right?

> 
> We'd need to account for this in our `ast_dump` implementation - which might be more trouble than it's worth...
> 

is reflecting on the enum hard?

> It might be interesting to see what the `SymbolTable` would look like if were a subclass of `ASTGenericVisitor`

yeah cool idea, but our testing for symtable is not good enough to refactor it I feel

> https://github.com/mozillazg/pypy/blob/master/pypy/interpreter/astcompiler/ast.py#L3661
> https://github.com/mozillazg/pypy/blob/master/pypy/interpreter/astcompiler/symtable.py#L312
> 
> Skulpt's compile implementation could also be a subclass of `ASTGenericVisitor` (maybe).
> 
> Should I move `ASTVisitor` to its own file?

that makes sense to me. does it have to be generated?

> if not then is the `astnodes.ts` the wrong name for the file...